src/dataflow/src/decode/csv.rs

```diff
@@ -9,193 +9,151 @@
+                            Err(DataflowError::DecodeError(DecodeError::Text(format!(
+                                "CSV error at lineno {}: expected {} columns, got {}.",
+                                self.total_events(),
+                                self.n_cols,
+                                ends_valid
+                            ))))
+                        } else {
+                            match std::str::from_utf8(&self.output[0..self.output_cursor]) {
+                                Ok(output) => {
+                                    self.events_success += 1;
+                                    let mut row_packer = std::mem::take(&mut self.row_packer);
+                                    if push_metadata {
+                                        row_packer.extend(
+                                            (0..self.n_cols)
+                                                .map(|i| {
+                                                    Datum::String(if self.demanded[i] {
+                                                        &output[self.ends[i]..self.ends[i + 1]]
+                                                    } else {
+                                                        ""
                                                     })
-                                                    .chain(iter::once(
-                                                        line_no.map(Datum::Int64).into(),
-                                                    )),
-                                            );
-                                            session.give((
-                                                Ok(row_packer.finish_and_reuse()),
-                                                *cap.time(),
-                                                1,
-                                            ));
-                                            // Reset valid data to extract the next record, should one exist.
-                                            buffer_valid = 0;
-                                            bounds_valid = 0;
-                                        }
-                                    }
-                                    csv_core::ReadRecordResult::End => {
-                                        done = true;
+                                                })
+                                                .chain(iter::once(Datum::from(coord))),
+                                        );
+                                    } else {
+                                        row_packer.extend((0..self.n_cols).map(|i| {
+                                            Datum::String(if self.demanded[i] {
+                                                &output[self.ends[i]..self.ends[i + 1]]
+                                            } else {
+                                                ""
+                                            })
+                                        }));
```

Comment:
Yes, but I wanted to avoid the extra memory allocation caused by the `row_packer.push` .

Using `row_packer.extend` in both cases should let the row be allocated to its correct size in one shot.